src/linear_algebra/trace.lean

```diff
@@ -115,6 +116,28 @@ end
 section
 variables (R : Type u) [field R] {M : Type v} [add_comm_group M] [module R M]
 
+/-- The trace of a linear map correspond to the contraction pairing under the isomorphism
+ `M →ₗ M ≃ M* ⊗ M`-/
+lemma trace_eq_contract [finite_dimensional R M] :
+  (linear_map.trace R M) ∘ₗ ↑(dual_tensor_hom_equiv R M M) = contract_left R M :=
+begin
+  have b := fin_basis R M,
+  apply basis.ext (basis.tensor_product (b.dual_basis) b),
+  rintros ⟨i, j⟩,
+  simp [dual_tensor_hom_equiv_to_lin],
```

Comment:
A call to `simp` that does not close the goal makes for brittle proofs since `simp` is sensitive to global state. For this reason we never allow such so-called "non-terminal `simp`s". The easiest way out of a situation like this is often to use `squeeze_simp` to turn your non-terminal `simp` into a `simp only` (which is allowed mid-proof). In this case we get:
```lean
  simp only [dual_tensor_hom_equiv_to_lin, basis.coord_apply, linear_equiv.coe_coe,
    basis.repr_self, function.comp_app, contract_left_apply, basis.tensor_product_apply,
    basis.coe_dual_basis, coe_comp],
```
An alternative is to specify the desired new goal state using `suffices` and keep your `simp`.

Reply:
Changed to simp_only